io-functions-pay-portal/PostNewslettersRecipients/handler.ts

```diff
@@ -0,0 +1,276 @@
+  context: Context,
+  id: NonEmptyString,
+  recipientRequest: RecipientRequest
+) => Promise<IResponseSuccessJson<RecipientResponse> | ErrorResponses>;
+
+/**
+ * Model for mailup OAuth token
+ */
+const AuthTokenR = t.partial({});
+
+const AuthTokenO = t.interface({
+  access_token: NonEmptyString,
+
+  expires_in: t.number,
+
+  refresh_token: t.string
+});
+
+const MailupAuthToken = t.intersection(
+  [AuthTokenR, AuthTokenO],
+  "MailupAuthToken"
+);
+
+type MailupAuthToken = t.TypeOf<typeof MailupAuthToken>;
+
+/**
+ * Model for Google reCaptcha response
+ */
+const ResponseR = t.interface({
+  success: t.boolean,
+
+  challenge_ts: t.string,
+
+  hostname: t.string
+});
+
+const ResponseO = t.partial({
+  "error-codes": t.readonlyArray(t.string, "array of string")
+});
+
+export const ResponseRecaptcha = t.intersection(
+  [ResponseR, ResponseO],
+  "ResponseRecaptcha"
+);
+
+export type ResponseRecaptcha = t.TypeOf<typeof ResponseRecaptcha>;
+
+/**
+ * A middleware that extracts a RecipientRequest payload from a request.
+ */
+const RecipientRequestMiddleware: IRequestMiddleware<
+  "IResponseErrorValidation",
+  RecipientRequest
+> = request =>
+  Promise.resolve(
+    RecipientRequest.decode(request.body).mapLeft(
+      ResponseErrorFromValidationErrors(RecipientRequest)
+    )
+  );
+
+const recaptchaCheckTask = (
+  recaptchaToken: string
+): TaskEither<Error, ResponseRecaptcha> =>
+  tryCatch(
+    () =>
+      fetchApi(`https://www.google.com/recaptcha/api/siteverify`, {
+        body: `secret=${config.RECAPTCHA_SECRET}&response=${recaptchaToken}`,
+        headers: {
+          // tslint:disable-next-line: no-duplicate-string
+          "Content-Type": "application/x-www-form-urlencoded"
+        },
+        method: "POST"
+      }),
+    err => new Error(`Error posting recaptcha API: ${err}`)
+  )
+    .chain(
+      fromPredicate<Error, Response>(
+        r => r.ok,
+        r => new Error(`Error returned from recaptcha API: ${r.status}`)
+      )
+    )
+    .chain(response =>
+      tryCatch(
+        () => response.json(),
+        err => new Error(`Error getting recaptcha API payload: ${err}`)
+      )
+    )
+    .chain(json =>
+      fromEither(
+        ResponseRecaptcha.decode(json).mapLeft(
+          errors =>
+            new Error(
+              `Error while decoding response from recaptcha: ${readableReport(
+                errors
+              )})`
+            )
+        )
+      )
+    )
+    .chain(
+      fromPredicate(
+        ar => ar.success,
+        _ => new Error(`Error checking recaptcha`)
+      )
+    );
+
+const getMailupAuthTokenTask = (): TaskEither<Error, MailupAuthToken> =>
+  tryCatch(
+    () =>
+      fetchApi(`https://services.mailup.com/Authorization/OAuth/Token`, {
+        body: `grant_type=password&client_id=${config.MAILUP_CLIENT_ID}&client_secret=${config.MAILUP_SECRET}&username=${config.MAILUP_USERNAME}&password=${config.MAILUP_PASSWORD}`,
```

Comment:
I'm unconvinced, why a Resource Owner flow?
I would have probably picked a Client Credentials flow, I find it difficult to distinguish between a Client and a Resource Owner here.

Does a "grant_type=client_credentials" work, after you remove "username" and "password"?

Reply:
@pp-ps  Unfortunately, the _Authorization code grant_ described in the [documentation](https://help.mailup.com/display/mailupapi/Authenticating+with+OAuth+v2) cannot be used in this case (rest API), as it requires entering user and password on the _mailup_ page and then defining a callback url. In the _Resource Owner Password Grant_ described by the [documentation](https://help.mailup.com/display/mailupapi/Authenticating+with+OAuth+v2) , unfortunately, user and password is mandatory, however a client id is also passed that allows you to distinguish the  _Client_ and a _Resource Owner_.
This kind of authentication _grant_type=client_credentials_ it is not documented, but I can try to do an empirical test.